File: cassandra/src/main/java/com/dbx/agent/cassandra/CassandraAgent.java
package com.dbx.agent.cassandra;

import com.dbx.agent.AbstractJdbcAgent;
import com.dbx.agent.ColumnInfo;
import com.dbx.agent.ConnectParams;
import com.dbx.agent.DatabaseInfo;
import com.dbx.agent.ForeignKeyInfo;
import com.dbx.agent.IndexInfo;
import com.dbx.agent.JdbcIdentifiers;
import com.dbx.agent.JsonRpcServer;
import com.dbx.agent.TableInfo;
import com.dbx.agent.TriggerInfo;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class CassandraAgent extends AbstractJdbcAgent {
    private static final Pattern TARGET_PATTERN = Pattern.compile("target[\"']?\\s*[:=]\\s*[\"']?([\\w]+)");

    @Override
    protected String driverClass() {
        return "com.ing.data.cassandra.jdbc.CassandraDriver";
    }

    @Override
    protected String buildJdbcUrl(ConnectParams params) {
        return buildUrl(params);
    }

    @Override
    public String setSchemaSQL(String schema) {
        return "USE " + JdbcIdentifiers.INSTANCE.doubleQuote(schema);
    }

    @Override
    public List<DatabaseInfo> listDatabases() {
        return unchecked(() -> {
            List<DatabaseInfo> result = new ArrayList<>();
            String sql = "SELECT keyspace_name FROM system_schema.keyspaces";
            try (java.sql.Statement stmt = requireConnected().createStatement();
                 ResultSet rs = stmt.executeQuery(sql)) {
                while (rs.next()) {
                    result.add(new DatabaseInfo(rs.getString(1)));
                }
            }
            result.sort(Comparator.comparing(DatabaseInfo::getName));
            return result;
        });
    }

    @Override
    public List<String> listSchemas() {
        List<String> result = new ArrayList<>();
        for (DatabaseInfo database : listDatabases()) {
            result.add(database.getName());
        }
        return result;
    }

    @Override
    public List<TableInfo> listTables(String schema) {
        return unchecked(() -> {
            List<TableInfo> result = new ArrayList<>();
            String sql = "SELECT table_name FROM system_schema.tables WHERE keyspace_name = ?";
            try (PreparedStatement stmt = requireConnected().prepareStatement(sql)) {
                stmt.setString(1, schema);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        result.add(new TableInfo(rs.getString(1), "TABLE", null));
                    }
                }
            }
            result.sort(Comparator.comparing(TableInfo::getName));
            return result;
        });
    }

    @Override
    public List<ColumnInfo> getColumns(String schema, String table) {
        return unchecked(() -> {
            List<ColumnInfo> result = new ArrayList<>();
            String sql = "SELECT column_name, type, kind FROM system_schema.columns WHERE keyspace_name = ? AND table_name = ?";
            try (PreparedStatement stmt = requireConnected().prepareStatement(sql)) {
                stmt.setString(1, schema);
                stmt.setString(2, table);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        String kind = coalesce(rs.getString("kind"));
                        boolean isPrimaryKey = "partition_key".equals(kind) || "clustering".equals(kind);
                        result.add(new ColumnInfo(
                            rs.getString("column_name"),
                            coalesce(rs.getString("type"), "unknown"),
                            !isPrimaryKey,
                            null,
                            isPrimaryKey,
                            kind.trim().isEmpty() ? null : kind,
                            null,
                            null,
                            null,
                            null
                        ));
                    }
                }
            }
            return result;
        });
    }

    @Override
    public List<IndexInfo> listIndexes(String schema, String table) {
        return unchecked(() -> {
            List<IndexInfo> result = new ArrayList<>();
            String sql = "SELECT index_name, options FROM system_schema.indexes WHERE keyspace_name = ? AND table_name = ?";
            try (PreparedStatement stmt = requireConnected().prepareStatement(sql)) {
                stmt.setString(1, schema);
                stmt.setString(2, table);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        String indexName = coalesce(rs.getString("index_name"));
                        String options = coalesce(rs.getString("options"));
                        result.add(new IndexInfo(
                            indexName,
                            targetColumns(options),
                            false,
                            false,
                            null,
                            null,
                            null,
                            null
                        ));
                    }
                }
            }
            result.sort(Comparator.comparing(IndexInfo::getName));
            return result;
        });
    }

    @Override
    public List<ForeignKeyInfo> listForeignKeys(String schema, String table) {
        return Collections.emptyList();
    }

    @Override
    public List<TriggerInfo> listTriggers(String schema, String table) {
        return Collections.emptyList();
    }

    @Override
    protected Object resultValue(ResultSet rs, int index, int sqlType) {
        return unchecked(() -> {
            Object value = rs.getObject(index);
            return rs.wasNull() ? null : value == null ? null : value.toString();
        });
    }

    private static String buildUrl(ConnectParams params) {
        return "jdbc:cassandra://" + params.getHost() + ":" + params.getPort() + "/" + params.getDatabase();
    }

    private static List<String> targetColumns(String options) {
        Matcher matcher = TARGET_PATTERN.matcher(options);
        if (!matcher.find()) {
            return Collections.emptyList();
        }
        return Collections.singletonList(matcher.group(1));
    }

    private static String coalesce(String value) {
        return value == null ? "" : value;
    }

    private static String coalesce(String value, String fallback) {
        return value == null ? fallback : value;
    }

    public static void main(String[] args) {
        new JsonRpcServer(new CassandraAgent()).run();
    }
}

File: cassandra/src/main/kotlin/com/dbx/agent/cassandra/CassandraAgent.kt
package com.dbx.agent.cassandra

import com.dbx.agent.*
import java.sql.Connection
import java.sql.DriverManager

class CassandraAgent : DatabaseAgent {

    private var connection: Connection? = null

    override fun getConnection(): Connection? = connection

    override fun setSchemaSQL(schema: String): String = "USE ${JdbcIdentifiers.doubleQuote(schema)}"

    override fun connect(params: ConnectParams) {
        Class.forName("com.ing.data.cassandra.jdbc.CassandraDriver")
        val url = "jdbc:cassandra://${params.host}:${params.port}/${params.database}"
        connection = DriverManager.getConnection(url, params.username, params.password)
    }

    override fun testConnection(params: ConnectParams): Boolean {
        Class.forName("com.ing.data.cassandra.jdbc.CassandraDriver")
        val url = "jdbc:cassandra://${params.host}:${params.port}/${params.database}"
        DriverManager.getConnection(url, params.username, params.password).use { conn ->
            return conn.isValid(5)
        }
    }

    override fun listDatabases(): List<DatabaseInfo> {
        val conn = requireConnection()
        val sql = "SELECT keyspace_name FROM system_schema.keyspaces"
        conn.createStatement().use { stmt ->
            stmt.executeQuery(sql).use { rs ->
                return buildList {
                    while (rs.next()) {
                        add(DatabaseInfo(rs.getString(1)))
                    }
                }.sortedBy { it.name }
            }
        }
    }

    override fun listSchemas(): List<String> {
        return listDatabases().map { it.name }
    }

    override fun listTables(schema: String): List<TableInfo> {
        val conn = requireConnection()
        val sql = "SELECT table_name FROM system_schema.tables WHERE keyspace_name = ?"
        conn.prepareStatement(sql).use { stmt ->
            stmt.setString(1, schema)
            stmt.executeQuery().use { rs ->
                return buildList {
                    while (rs.next()) {
                        add(TableInfo(
                            name = rs.getString(1),
                            table_type = "TABLE"
                        ))
                    }
                }.sortedBy { it.name }
            }
        }
    }

    override fun getColumns(schema: String, table: String): List<ColumnInfo> {
        val conn = requireConnection()
        val sql = "SELECT column_name, type, kind FROM system_schema.columns WHERE keyspace_name = ? AND table_name = ?"
        conn.prepareStatement(sql).use { stmt ->
            stmt.setString(1, schema)
            stmt.setString(2, table)
            stmt.executeQuery().use { rs ->
                return buildList {
                    while (rs.next()) {
                        val kind = rs.getString("kind") ?: ""
                        val isPrimaryKey = kind == "partition_key" || kind == "clustering"
                        add(ColumnInfo(
                            name = rs.getString("column_name"),
                            data_type = rs.getString("type") ?: "unknown",
                            is_nullable = !isPrimaryKey,
                            column_default = null,
                            is_primary_key = isPrimaryKey,
                            extra = if (kind.isNotBlank()) kind else null,
                            comment = null
                        ))
                    }
                }
            }
        }
    }

    override fun listIndexes(schema: String, table: String): List<IndexInfo> {
        val conn = requireConnection()
        val sql = "SELECT index_name, options FROM system_schema.indexes WHERE keyspace_name = ? AND table_name = ?"
        conn.prepareStatement(sql).use { stmt ->
            stmt.setString(1, schema)
            stmt.setString(2, table)
            stmt.executeQuery().use { rs ->
                return buildList {
                    while (rs.next()) {
                        val indexName = rs.getString("index_name") ?: ""
                        val options = rs.getString("options") ?: ""
                        // Extract target column from options map if possible
                        val targetMatch = Regex("target[\"']?\\s*[:=]\\s*[\"']?([\\w]+)").find(options)
                        val columns = if (targetMatch != null) listOf(targetMatch.groupValues[1]) else emptyList()
                        add(IndexInfo(
                            name = indexName,
                            columns = columns,
                            is_unique = false,
                            is_primary = false,
                            filter = null,
                            index_type = null
                        ))
                    }
                }.sortedBy { it.name }
            }
        }
    }

    override fun listForeignKeys(schema: String, table: String): List<ForeignKeyInfo> {
        return emptyList()
    }

    override fun listTriggers(schema: String, table: String): List<TriggerInfo> {
        return emptyList()
    }

    override fun executeQuery(sql: String, schema: String?): QueryResult {
        return JdbcExecutor.execute(requireConnection(), sql, schema, ::setSchemaSQL, valueReader = ::stringResultValue)
    }

    private fun stringResultValue(rs: java.sql.ResultSet, index: Int, sqlType: Int): Any? {
        val value = rs.getObject(index)
        return if (rs.wasNull()) null else value?.toString()
    }

    override fun disconnect() {
        connection?.close()
        connection = null
    }

    private fun requireConnection(): Connection {
        return connection ?: throw IllegalStateException("Not connected")
    }
}

fun main() {
    JsonRpcServer(CassandraAgent()).run()
}
